src/IO/ThreadPoolReader.h

```diff
@@ -0,0 +1,177 @@
+  * Note: doing reads from thread pool is usually bad idea for the following reasons:
+  * - for fast devices or when data is in page cache, it is less cache-friendly and less NUMA friendly
+  *   and also involves extra synchronization overhead;
+  * - for fast devices and lots of small random reads, it cannot utilize the device, because
+  *   OS will spent all the time in switching between threads and wasting CPU;
+  * - you don't know how many threads do you need, for example, when reading from page cache,
+  *   you need the number of threads similar to the number of CPU cores;
+  *   when reading from HDD array you need the number of threads as the number of disks times N;
+  *   when reading from SSD you may need at least hundreds of threads for efficient random reads,
+  *   but it is impractical;
+  * For example, this method is used in POSIX AIO that is notoriously useless (in contrast to Linux AIO).
+  *
+  * This is intended only as example for implementation of readers from remote filesystems,
+  * where this method can be feasible.
+  */
+class ThreadPoolReader final : public IAsynchronousReader
+{
+private:
+    ThreadPool pool;
+
+public:
+    ThreadPoolReader(size_t pool_size, size_t queue_size_)
+        : pool(pool_size, pool_size, queue_size_)
+    {
+    }
+
+    std::future<Result> submit(Request request) override
+    {
+        int fd = assert_cast<const LocalFileDescriptor &>(*request.descriptor).fd;
+
+#if defined(__linux__)
+        /// Check if data is already in page cache with preadv2 syscall.
+
+        /// TODO ProfileEvents for page cache hits and misses.
+
+        /// We don't want to depend on new Linux kernel.
+        static std::atomic<bool> has_pread_nowait_support{true};
+
+        if (has_pread_nowait_support.load(std::memory_order_relaxed))
+        {
+            std::promise<Result> promise;
+            std::future<Result> future = promise.get_future();
+
+            size_t bytes_read = 0;
+            while (!bytes_read)
+            {
+                struct iovec io_vec{ .iov_base = request.buf, .iov_len = request.size };
+                ssize_t res = syscall(
+                    SYS_preadv2, fd,
+                    &io_vec, 1,
+                    static_cast<long>(request.offset), static_cast<long>(request.offset >> 32),
+                    RWF_NOWAIT);
+
+                //ssize_t res = ::pread(fd, request.buf, request.size, request.offset);
+
+                if (!res)
+                {
+                    promise.set_value(0);
+                    return future;
+                }
+
+                if (-1 == res)
+                {
+                    if (errno == ENOSYS || errno == EOPNOTSUPP)
+                    {
+                        has_pread_nowait_support.store(false, std::memory_order_relaxed);
+                        break;
+                    }
+                    else if (errno == EAGAIN)
+                    {
+                        /// Data is not available.
```

Comment:
Not clear comment, maybe `Data is not available right now, need to do ordinary pread`.